<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Code &mdash; MDL 1 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Code - Data Management" href="code_data_mgmt.html" />
    <link rel="prev" title="Commands" href="commands_ngs.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> MDL
          </a>
              <div class="version">
                Version 1.4.5
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="ngs_pipeline.html">NGS pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="running_north.html">Running the pipeline at GLS North</a></li>
<li class="toctree-l1"><a class="reference internal" href="running_south.html">Running the pipeline at GLS South</a></li>
<li class="toctree-l1"><a class="reference internal" href="genes.html">Genes</a></li>
<li class="toctree-l1"><a class="reference internal" href="applications.html">Applications</a></li>
<li class="toctree-l1"><a class="reference internal" href="commands_ngs.html">Commands</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Code</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#code-from-the-pipeline-scripts">Code from the pipeline scripts</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#main-script">Main script</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#align-py">align.py</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#scripts-in-lib">Scripts in /lib</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#module-annovar">annovar.py</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-bwa_align">bwa_align.py</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-check_programs">check_programs.py</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-count2">count2.py</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-dup_indels">dup_indels.py</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-extract_identity">extract_identity.py</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-GATK_vcf">GATK_vcf.py</a></li>
<li class="toctree-l4"><a class="reference internal" href="#generate-final-report-py">generate_final_report.py</a></li>
<li class="toctree-l4"><a class="reference internal" href="#generate-full-final-report-py">generate_full_final_report.py</a></li>
<li class="toctree-l4"><a class="reference internal" href="#merge-tables-py">merge_tables.py</a></li>
<li class="toctree-l4"><a class="reference internal" href="#metrics-parser-py">metrics_parser.py</a></li>
<li class="toctree-l4"><a class="reference internal" href="#mpileup-py">mpileup.py</a></li>
<li class="toctree-l4"><a class="reference internal" href="#parse-annovar-py">parse_annovar.py</a></li>
<li class="toctree-l4"><a class="reference internal" href="#parse-interop-py">parse_interop.py</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-picard_actions">picard_actions.py</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-picard_metrics">picard_metrics.py</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-picard_parser">picard_parser.py</a></li>
<li class="toctree-l4"><a class="reference internal" href="#picard-qc-py">picard_qc.py</a></li>
<li class="toctree-l4"><a class="reference internal" href="#process-identity-py">process_identity.py</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-recalibration">recalibration.py</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-run_qualimap">run_qualimap.py</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-snpEff_ann">snpEff_ann.py</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-variants_freebayes">variants_freebayes.py</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-13">variants_octopus.py</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-variants_GATK">variants_GATK.py</a></li>
<li class="toctree-l4"><a class="reference internal" href="#variants-gatk3-py">variants_GATK3.py</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-18">variants_table.py</a></li>
<li class="toctree-l4"><a class="reference internal" href="#vcf-parser-py">vcf_parser.py</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="code_data_mgmt.html">Code - Data Management</a></li>
<li class="toctree-l1"><a class="reference internal" href="code_webapps.html">Code - Webapps</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">MDL</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
      <li>Code</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/code.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="code">
<h1>Code<a class="headerlink" href="#code" title="Permalink to this headline">¶</a></h1>
<section id="code-from-the-pipeline-scripts">
<h2>Code from the pipeline scripts<a class="headerlink" href="#code-from-the-pipeline-scripts" title="Permalink to this headline">¶</a></h2>
<section id="main-script">
<h3>Main script<a class="headerlink" href="#main-script" title="Permalink to this headline">¶</a></h3>
<section id="align-py">
<h4>align.py<a class="headerlink" href="#align-py" title="Permalink to this headline">¶</a></h4>
</section>
</section>
<section id="scripts-in-lib">
<h3>Scripts in /lib<a class="headerlink" href="#scripts-in-lib" title="Permalink to this headline">¶</a></h3>
<section id="module-annovar">
<span id="annovar-py"></span><h4>annovar.py<a class="headerlink" href="#module-annovar" title="Permalink to this headline">¶</a></h4>
<span class="target" id="module-0"></span><dl class="py function">
<dt class="sig sig-object py" id="annovar.get_annotation">
<span class="sig-prename descclassname"><span class="pre">annovar.</span></span><span class="sig-name descname"><span class="pre">get_annotation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">datadir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annovar</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/annovar.html#get_annotation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#annovar.get_annotation" title="Permalink to this definition">¶</a></dt>
<dd><p>Function that runs Annovar to annotate the merged VCF file</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_id</strong> (<em>string</em>) – ID of the sample being analysed</p></li>
<li><p><strong>datadir</strong> (<em>string</em>) – run location</p></li>
<li><p><strong>annovar</strong> (<em>string</em>) – location of the annovar executable (Perl)</p></li>
</ul>
</dd>
</dl>
<p>:todo improve return
:todo improve output
:todo parse and DB connection</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="annovar.get_code">
<span class="sig-prename descclassname"><span class="pre">annovar.</span></span><span class="sig-name descname"><span class="pre">get_code</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_id</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/annovar.html#get_code"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#annovar.get_code" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</section>
<section id="module-bwa_align">
<span id="bwa-align-py"></span><h4>bwa_align.py<a class="headerlink" href="#module-bwa_align" title="Permalink to this headline">¶</a></h4>
<span class="target" id="module-1"></span><dl class="py function">
<dt class="sig sig-object py" id="bwa_align.run_bwa">
<span class="sig-prename descclassname"><span class="pre">bwa_align.</span></span><span class="sig-name descname"><span class="pre">run_bwa</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fastq_files</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">datadir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bwa</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">samtools</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bwa_align.html#run_bwa"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bwa_align.run_bwa" title="Permalink to this definition">¶</a></dt>
<dd><p>function that performs the alignment/mapping, it is run iteratively for each pair/sample_id</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_id</strong> – </p></li>
<li><p><strong>fastq_files</strong> – </p></li>
<li><p><strong>datadir</strong> – </p></li>
<li><p><strong>reference</strong> – </p></li>
<li><p><strong>bwa</strong> – </p></li>
<li><p><strong>samtools</strong> – </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-check_programs">
<span id="check-programs-py"></span><h4>check_programs.py<a class="headerlink" href="#module-check_programs" title="Permalink to this headline">¶</a></h4>
<span class="target" id="module-2"></span><dl class="py function">
<dt class="sig sig-object py" id="check_programs.cmd_exists">
<span class="sig-prename descclassname"><span class="pre">check_programs.</span></span><span class="sig-name descname"><span class="pre">cmd_exists</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/check_programs.html#cmd_exists"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#check_programs.cmd_exists" title="Permalink to this definition">¶</a></dt>
<dd><p>Function that runs the defined command and returns its result</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>cmd</strong> (<em>string</em>) – Command to be tested</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the result of a subprocess.call operation with the desired command</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="check_programs.find_version">
<span class="sig-prename descclassname"><span class="pre">check_programs.</span></span><span class="sig-name descname"><span class="pre">find_version</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">stream</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/check_programs.html#find_version"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#check_programs.find_version" title="Permalink to this definition">¶</a></dt>
<dd><p>Function that tries to find the external program version from either stdout or sterr outputs</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>stream</strong> (<em>string</em>) – stdout and stderr combined in a single stream</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>returns the line containing the word version</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>string</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="check_programs.get_location">
<span class="sig-prename descclassname"><span class="pre">check_programs.</span></span><span class="sig-name descname"><span class="pre">get_location</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">software_conf</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/check_programs.html#get_location"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#check_programs.get_location" title="Permalink to this definition">¶</a></dt>
<dd><p>Function that checks the locations from the required external software based on a YAML
configuraiton file</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>software_conf</strong> – output from a YAML file stream with the key/value of required software</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>returns a dictionary with program:”found” or “error”</p>
</dd>
<dt class="field-odd">Todo</dt>
<dd class="field-odd"><p>make it easier for align.py to act/check this list</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="check_programs.get_version">
<span class="sig-prename descclassname"><span class="pre">check_programs.</span></span><span class="sig-name descname"><span class="pre">get_version</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/check_programs.html#get_version"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#check_programs.get_version" title="Permalink to this definition">¶</a></dt>
<dd><p>Function that tries to extract the version of external programs required for analysis</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>cmd</strong> (<em>string</em>) – Command to be checked</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>returns the standard output and standard error of a subprocess.Popen call</p>
</dd>
<dt class="field-odd">Todo</dt>
<dd class="field-odd"><p>log the return</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="check_programs.versions">
<span class="sig-prename descclassname"><span class="pre">check_programs.</span></span><span class="sig-name descname"><span class="pre">versions</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/check_programs.html#versions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#check_programs.versions" title="Permalink to this definition">¶</a></dt>
<dd><p>Main function of the script where the required external apps/languages are listed and called
No parameters</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>returns a dictionary with the program and its version</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-count2">
<span id="count2-py"></span><h4>count2.py<a class="headerlink" href="#module-count2" title="Permalink to this headline">¶</a></h4>
<span class="target" id="module-count2.py"></span><dl class="py function">
<dt class="sig sig-object py" id="count2.extract_counts">
<span class="sig-prename descclassname"><span class="pre">count2.</span></span><span class="sig-name descname"><span class="pre">extract_counts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">datadir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">full_BED</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_id</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/count2.html#extract_counts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#count2.extract_counts" title="Permalink to this definition">¶</a></dt>
<dd><p>Function that reads the BAM file and extract the read count for each window</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>datadir</strong> (<em>string</em>) – Location of the BAM files</p></li>
<li><p><strong>full_BED</strong> (<em>string</em>) – BED file to guide the counts</p></li>
<li><p><strong>sample_id</strong> (<em>string</em>) – ID of the sdample</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>no return yet</p>
</dd>
<dt class="field-odd">Todo</dt>
<dd class="field-odd"><p>change BED file location from hardcoded</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-dup_indels">
<span id="dup-indels-py"></span><h4>dup_indels.py<a class="headerlink" href="#module-dup_indels" title="Permalink to this headline">¶</a></h4>
<span class="target" id="module-3"></span><dl class="py function">
<dt class="sig sig-object py" id="dup_indels.add_groups">
<span class="sig-prename descclassname"><span class="pre">dup_indels.</span></span><span class="sig-name descname"><span class="pre">add_groups</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">datadir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">picard</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">samtools</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dup_indels.html#add_groups"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dup_indels.add_groups" title="Permalink to this definition">¶</a></dt>
<dd><p>Function that add header groups to BAM files generated by BWA</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_id</strong> (<em>string</em>) – ID of the patient/sample being analysed using Picard</p></li>
<li><p><strong>datadir</strong> (<em>string</em>) – Location of the BAM files</p></li>
<li><p><strong>samtools</strong> (<em>string</em>) – Location of the Samtools executable</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>returns success or exists</p>
</dd>
<dt class="field-odd">Todo</dt>
<dd class="field-odd"><p>return error</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="dup_indels.remove_duplicates">
<span class="sig-prename descclassname"><span class="pre">dup_indels.</span></span><span class="sig-name descname"><span class="pre">remove_duplicates</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">datadir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">picard</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dup_indels.html#remove_duplicates"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dup_indels.remove_duplicates" title="Permalink to this definition">¶</a></dt>
<dd><p>Function that runs the duplicate removal step in the analysis</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_id</strong> (<em>string</em>) – ID of the patient/sample being analysed using Picard</p></li>
<li><p><strong>datadir</strong> (<em>string</em>) – Location of the BAM files</p></li>
<li><p><strong>picard</strong> (<em>string</em>) – Picard jar file location</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>returns success or exists</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="dup_indels.samtools_duplicates">
<span class="sig-prename descclassname"><span class="pre">dup_indels.</span></span><span class="sig-name descname"><span class="pre">samtools_duplicates</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">datadir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">samtools</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dup_indels.html#samtools_duplicates"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dup_indels.samtools_duplicates" title="Permalink to this definition">¶</a></dt>
<dd><p>Function that runs the duplicate removal step in the analysis</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_id</strong> (<em>string</em>) – ID of the patient/sample being analysed using Picard</p></li>
<li><p><strong>datadir</strong> (<em>string</em>) – Location of the BAM files</p></li>
<li><p><strong>samtools</strong> (<em>string</em>) – Picard jar file location</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>returns success or exists</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-extract_identity">
<span id="extract-identity-py"></span><h4>extract_identity.py<a class="headerlink" href="#module-extract_identity" title="Permalink to this headline">¶</a></h4>
<span class="target" id="module-4"></span><dl class="py function">
<dt class="sig sig-object py" id="extract_identity.create_identity_table">
<span class="sig-prename descclassname"><span class="pre">extract_identity.</span></span><span class="sig-name descname"><span class="pre">create_identity_table</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">datadir</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/extract_identity.html#create_identity_table"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#extract_identity.create_identity_table" title="Permalink to this definition">¶</a></dt>
<dd><p>Function that creates the identity table in the sample
datadir</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_id</strong> (<em>string</em>) – patient id to be analyses</p></li>
<li><p><strong>datadir</strong> (<em>string</em>) – datadir where the table will be saved</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="extract_identity.get_nucleotides">
<span class="sig-prename descclassname"><span class="pre">extract_identity.</span></span><span class="sig-name descname"><span class="pre">get_nucleotides</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pileup</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/extract_identity.html#get_nucleotides"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#extract_identity.get_nucleotides" title="Permalink to this definition">¶</a></dt>
<dd><p>Function that returns nucleotide counts for the region</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>pileup</strong> (<em>string</em>) – reads under a defined location</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>returns the number of A, C, G and T</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="extract_identity.mpileup">
<span class="sig-prename descclassname"><span class="pre">extract_identity.</span></span><span class="sig-name descname"><span class="pre">mpileup</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">datadir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">identity</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">samtools</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/extract_identity.html#mpileup"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#extract_identity.mpileup" title="Permalink to this definition">¶</a></dt>
<dd><p>Function that call sammtool to generate the pileup
file that can be searched for the nucleotide state in
the locations of interest</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_id</strong> (<em>string</em>) – ID of the patient/sample being analysed</p></li>
<li><p><strong>datadir</strong> (<em>string</em>) – Location of the BAM files</p></li>
<li><p><strong>identity</strong> (<em>string</em>) – Indentity file name with defined location</p></li>
<li><p><strong>samtools</strong> (<em>string</em>) – Samtools executable location</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>returns success or exists</p>
</dd>
<dt class="field-odd">Todo</dt>
<dd class="field-odd"><p>return error</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-GATK_vcf">
<span id="gatk-vcf-py"></span><h4>GATK_vcf.py<a class="headerlink" href="#module-GATK_vcf" title="Permalink to this headline">¶</a></h4>
<span class="target" id="module-5"></span><dl class="py function">
<dt class="sig sig-object py" id="GATK_vcf.vcf_comparison">
<span class="sig-prename descclassname"><span class="pre">GATK_vcf.</span></span><span class="sig-name descname"><span class="pre">vcf_comparison</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">datadir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gatk</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/GATK_vcf.html#vcf_comparison"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#GATK_vcf.vcf_comparison" title="Permalink to this definition">¶</a></dt>
<dd><p>Function that merges the available VCFs in the sample VCF datadir</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_id</strong> (<em>string</em>) – ID of the patient/sample being analysed using GATK</p></li>
<li><p><strong>datadir</strong> (<em>string</em>) – Location of the BAM files</p></li>
<li><p><strong>bed_file</strong> (<em>string</em>) – BED file with regions to be analysed</p></li>
<li><p><strong>reference</strong> (<em>string</em>) – Reference file used in the original alignment</p></li>
<li><p><strong>gatk</strong> (<em>string</em>) – GATK jar file location</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>returns success or exists</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="generate-final-report-py">
<h4>generate_final_report.py<a class="headerlink" href="#generate-final-report-py" title="Permalink to this headline">¶</a></h4>
</section>
<section id="generate-full-final-report-py">
<h4>generate_full_final_report.py<a class="headerlink" href="#generate-full-final-report-py" title="Permalink to this headline">¶</a></h4>
</section>
<section id="merge-tables-py">
<h4>merge_tables.py<a class="headerlink" href="#merge-tables-py" title="Permalink to this headline">¶</a></h4>
</section>
<section id="metrics-parser-py">
<h4>metrics_parser.py<a class="headerlink" href="#metrics-parser-py" title="Permalink to this headline">¶</a></h4>
</section>
<section id="mpileup-py">
<h4>mpileup.py<a class="headerlink" href="#mpileup-py" title="Permalink to this headline">¶</a></h4>
</section>
<section id="parse-annovar-py">
<h4>parse_annovar.py<a class="headerlink" href="#parse-annovar-py" title="Permalink to this headline">¶</a></h4>
</section>
<section id="parse-interop-py">
<h4>parse_interop.py<a class="headerlink" href="#parse-interop-py" title="Permalink to this headline">¶</a></h4>
</section>
<section id="module-picard_actions">
<span id="picard-actions-py"></span><h4>picard_actions.py<a class="headerlink" href="#module-picard_actions" title="Permalink to this headline">¶</a></h4>
<span class="target" id="module-6"></span><dl class="py function">
<dt class="sig sig-object py" id="picard_actions.picard_sort">
<span class="sig-prename descclassname"><span class="pre">picard_actions.</span></span><span class="sig-name descname"><span class="pre">picard_sort</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">datadir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">picard</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/picard_actions.html#picard_sort"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#picard_actions.picard_sort" title="Permalink to this definition">¶</a></dt>
<dd><p>Function that calls Picard for Freebayes VCF sorting</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_id</strong> (<em>string</em>) – ID of the patient/sample being analysed</p></li>
<li><p><strong>datadir</strong> (<em>string</em>) – Location of the BAM files</p></li>
<li><p><strong>reference</strong> (<em>string</em>) – Reference file used in the original alignment</p></li>
<li><p><strong>picard</strong> (<em>string</em>) – Locations of the picard jar file</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-picard_metrics">
<span id="picard-metrics-py"></span><h4>picard_metrics.py<a class="headerlink" href="#module-picard_metrics" title="Permalink to this headline">¶</a></h4>
<span class="target" id="module-7"></span><dl class="py function">
<dt class="sig sig-object py" id="picard_metrics.get_align_summary">
<span class="sig-prename descclassname"><span class="pre">picard_metrics.</span></span><span class="sig-name descname"><span class="pre">get_align_summary</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">directory</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">picard</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/picard_metrics.html#get_align_summary"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#picard_metrics.get_align_summary" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_id</strong> (<em>string</em>) – ID of the patient/sample being analysed using Picard</p></li>
<li><p><strong>directory</strong> (<em>string</em>) – Location of the BAM files</p></li>
<li><p><strong>reference</strong> (<em>string</em>) – Reference genome</p></li>
<li><p><strong>bait_file</strong> (<em>string</em>) – Picard specific BED file</p></li>
<li><p><strong>picard</strong> (<em>string</em>) – Picard jar file location</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>returns success or exists</p>
</dd>
<dt class="field-odd">Todo</dt>
<dd class="field-odd"><p>return error</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="picard_metrics.get_align_summary_parp">
<span class="sig-prename descclassname"><span class="pre">picard_metrics.</span></span><span class="sig-name descname"><span class="pre">get_align_summary_parp</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">directory</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">picard</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/picard_metrics.html#get_align_summary_parp"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#picard_metrics.get_align_summary_parp" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_id</strong> (<em>string</em>) – ID of the patient/sample being analysed using Picard</p></li>
<li><p><strong>directory</strong> (<em>string</em>) – Location of the BAM files</p></li>
<li><p><strong>reference</strong> (<em>string</em>) – Reference genome</p></li>
<li><p><strong>bait_file</strong> (<em>string</em>) – Picard specific BED file</p></li>
<li><p><strong>picard</strong> (<em>string</em>) – Picard jar file location</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>returns success or exists</p>
</dd>
<dt class="field-odd">Todo</dt>
<dd class="field-odd"><p>return error</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="picard_metrics.get_call_metrics">
<span class="sig-prename descclassname"><span class="pre">picard_metrics.</span></span><span class="sig-name descname"><span class="pre">get_call_metrics</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">directory</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vcf_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">picard</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/picard_metrics.html#get_call_metrics"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#picard_metrics.get_call_metrics" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_id</strong> (<em>string</em>) – ID of the patient/sample being analysed using Picard</p></li>
<li><p><strong>directory</strong> (<em>string</em>) – Location of the BAM files</p></li>
<li><p><strong>reference</strong> (<em>string</em>) – Reference genome</p></li>
<li><p><strong>bait_file</strong> (<em>string</em>) – Picard specific BED file</p></li>
<li><p><strong>picard</strong> (<em>string</em>) – Picard jar file location</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>returns success or exists</p>
</dd>
<dt class="field-odd">Todo</dt>
<dd class="field-odd"><p>return error</p>
</dd>
<dt class="field-even">Todo</dt>
<dd class="field-even"><p>fix argument</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="picard_metrics.get_hs_metrics">
<span class="sig-prename descclassname"><span class="pre">picard_metrics.</span></span><span class="sig-name descname"><span class="pre">get_hs_metrics</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">directory</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bait_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">picard</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">panel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'full'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/picard_metrics.html#get_hs_metrics"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#picard_metrics.get_hs_metrics" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_id</strong> (<em>string</em>) – ID of the patient/sample being analysed using Picard</p></li>
<li><p><strong>directory</strong> (<em>string</em>) – Location of the BAM files</p></li>
<li><p><strong>reference</strong> (<em>string</em>) – Reference genome</p></li>
<li><p><strong>bait_file</strong> (<em>string</em>) – Picard specific BED file</p></li>
<li><p><strong>picard</strong> (<em>string</em>) – Picard jar file location</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>returns success or exists</p>
</dd>
<dt class="field-odd">Todo</dt>
<dd class="field-odd"><p>return error</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="picard_metrics.get_hs_metrics_parp">
<span class="sig-prename descclassname"><span class="pre">picard_metrics.</span></span><span class="sig-name descname"><span class="pre">get_hs_metrics_parp</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">directory</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bait_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">picard</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/picard_metrics.html#get_hs_metrics_parp"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#picard_metrics.get_hs_metrics_parp" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_id</strong> (<em>string</em>) – ID of the patient/sample being analysed using Picard</p></li>
<li><p><strong>directory</strong> (<em>string</em>) – Location of the BAM files</p></li>
<li><p><strong>reference</strong> (<em>string</em>) – Reference genome</p></li>
<li><p><strong>bait_file</strong> (<em>string</em>) – Picard specific BED file</p></li>
<li><p><strong>picard</strong> (<em>string</em>) – Picard jar file location</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>returns success or exists</p>
</dd>
<dt class="field-odd">Todo</dt>
<dd class="field-odd"><p>return error</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="picard_metrics.get_yield">
<span class="sig-prename descclassname"><span class="pre">picard_metrics.</span></span><span class="sig-name descname"><span class="pre">get_yield</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">directory</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">picard</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/picard_metrics.html#get_yield"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#picard_metrics.get_yield" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_id</strong> (<em>string</em>) – ID of the patient/sample being analysed using Picard</p></li>
<li><p><strong>directory</strong> (<em>string</em>) – Location of the BAM files</p></li>
<li><p><strong>reference</strong> (<em>string</em>) – Reference genome</p></li>
<li><p><strong>bait_file</strong> (<em>string</em>) – Picard specific BED file</p></li>
<li><p><strong>picard</strong> (<em>string</em>) – Picard jar file location</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>returns success or exists</p>
</dd>
<dt class="field-odd">Todo</dt>
<dd class="field-odd"><p>return error</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="picard_metrics.get_yield_parp">
<span class="sig-prename descclassname"><span class="pre">picard_metrics.</span></span><span class="sig-name descname"><span class="pre">get_yield_parp</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">directory</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">picard</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/picard_metrics.html#get_yield_parp"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#picard_metrics.get_yield_parp" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_id</strong> (<em>string</em>) – ID of the patient/sample being analysed using Picard</p></li>
<li><p><strong>directory</strong> (<em>string</em>) – Location of the BAM files</p></li>
<li><p><strong>reference</strong> (<em>string</em>) – Reference genome</p></li>
<li><p><strong>bait_file</strong> (<em>string</em>) – Picard specific BED file</p></li>
<li><p><strong>picard</strong> (<em>string</em>) – Picard jar file location</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>returns success or exists</p>
</dd>
<dt class="field-odd">Todo</dt>
<dd class="field-odd"><p>return error</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-picard_parser">
<span id="picard-parser-py"></span><h4>picard_parser.py<a class="headerlink" href="#module-picard_parser" title="Permalink to this headline">¶</a></h4>
<span class="target" id="module-8"></span><dl class="py function">
<dt class="sig sig-object py" id="picard_parser.main_parser">
<span class="sig-prename descclassname"><span class="pre">picard_parser.</span></span><span class="sig-name descname"><span class="pre">main_parser</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">picard_output</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/picard_parser.html#main_parser"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#picard_parser.main_parser" title="Permalink to this definition">¶</a></dt>
<dd><p>Simple function that parses usual picard output files</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>picard_output</strong> (<em>string</em>) – location of the picard file</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>file contents</p>
</dd>
<dt class="field-odd">Todo</dt>
<dd class="field-odd"><p>return error</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="picard-qc-py">
<h4>picard_qc.py<a class="headerlink" href="#picard-qc-py" title="Permalink to this headline">¶</a></h4>
</section>
<section id="process-identity-py">
<h4>process_identity.py<a class="headerlink" href="#process-identity-py" title="Permalink to this headline">¶</a></h4>
</section>
<section id="module-recalibration">
<span id="recalibration-py"></span><h4>recalibration.py<a class="headerlink" href="#module-recalibration" title="Permalink to this headline">¶</a></h4>
<span class="target" id="module-9"></span><dl class="py function">
<dt class="sig sig-object py" id="recalibration.base_recal1">
<span class="sig-prename descclassname"><span class="pre">recalibration.</span></span><span class="sig-name descname"><span class="pre">base_recal1</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">directory</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bed_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vcf_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gatk</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/recalibration.html#base_recal1"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#recalibration.base_recal1" title="Permalink to this definition">¶</a></dt>
<dd><p>Function that does the first step of base recalibration, creating a
recalibration data table</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_id</strong> (<em>string</em>) – ID of the patient/sample being analysed</p></li>
<li><p><strong>directory</strong> (<em>string</em>) – Location of the BAM files</p></li>
<li><p><strong>bed_file</strong> (<em>string</em>) – BED file with regions to be analysed</p></li>
<li><p><strong>vcf_file</strong> (<em>string</em>) – VCF file of known regions of variants</p></li>
<li><p><strong>reference</strong> (<em>string</em>) – Reference file used in the original alignment</p></li>
<li><p><strong>gatk</strong> (<em>string</em>) – GATK jar file location</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>returns success or exists</p>
</dd>
<dt class="field-odd">Todo</dt>
<dd class="field-odd"><p>return error</p>
</dd>
<dt class="field-even">Todo</dt>
<dd class="field-even"><p>fix argument</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="recalibration.base_recal2">
<span class="sig-prename descclassname"><span class="pre">recalibration.</span></span><span class="sig-name descname"><span class="pre">base_recal2</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">directory</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vcf_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gatk</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/recalibration.html#base_recal2"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#recalibration.base_recal2" title="Permalink to this definition">¶</a></dt>
<dd><p>Function that does the second step of the base recalibration
process, generating the post-recalibration table</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_id</strong> (<em>string</em>) – ID of the patient/sample being analysed using GATK</p></li>
<li><p><strong>directory</strong> (<em>string</em>) – Location of the BAM files</p></li>
<li><p><strong>vcf_file</strong> (<em>string</em>) – VCF file of known regions of variants</p></li>
<li><p><strong>reference</strong> (<em>string</em>) – Reference file used in the original alignment</p></li>
<li><p><strong>gatk</strong> (<em>string</em>) – GATK jar file location</p></li>
</ul>
</dd>
<dt class="field-even">Todo</dt>
<dd class="field-even"><p>fix argument</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="recalibration.recalibrate">
<span class="sig-prename descclassname"><span class="pre">recalibration.</span></span><span class="sig-name descname"><span class="pre">recalibrate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">directory</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gatk</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/recalibration.html#recalibrate"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#recalibration.recalibrate" title="Permalink to this definition">¶</a></dt>
<dd><p>Function that performs the third step of the base recalibration process,
generating the final BAM file. <a href="#id1"><span class="problematic" id="id2">*</span></a>.recal_reads.bam</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_id</strong> (<em>string</em>) – ID of the patient/sample being analysed using GATK</p></li>
<li><p><strong>directory</strong> (<em>string</em>) – Location of the BAM files</p></li>
<li><p><strong>reference</strong> (<em>string</em>) – Reference file used in the original alignment</p></li>
<li><p><strong>gatk</strong> (<em>string</em>) – GATK jar file location</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-run_qualimap">
<span id="run-qualimap-py"></span><h4>run_qualimap.py<a class="headerlink" href="#module-run_qualimap" title="Permalink to this headline">¶</a></h4>
<span class="target" id="module-10"></span><dl class="py function">
<dt class="sig sig-object py" id="run_qualimap.call_qualimap">
<span class="sig-prename descclassname"><span class="pre">run_qualimap.</span></span><span class="sig-name descname"><span class="pre">call_qualimap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">directory</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bed_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">qualimap</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/run_qualimap.html#call_qualimap"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#run_qualimap.call_qualimap" title="Permalink to this definition">¶</a></dt>
<dd><p>Function that performs the call to qualimap</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_id</strong> (<em>string</em>) – ID of the patient/sample being analysed</p></li>
<li><p><strong>directory</strong> (<em>string</em>) – Location of the BAM files</p></li>
<li><p><strong>bed_file</strong> (<em>string</em>) – BED file with regions to be analysed</p></li>
<li><p><strong>qualimap</strong> (<em>string</em>) – Qualimap jar/bat/etc location</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>returns success or exists</p>
</dd>
<dt class="field-odd">Todo</dt>
<dd class="field-odd"><p>return error</p>
</dd>
<dt class="field-even">Todo</dt>
<dd class="field-even"><p>fix argument</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-snpEff_ann">
<span id="snpeff-ann-py"></span><h4>snpEff_ann.py<a class="headerlink" href="#module-snpEff_ann" title="Permalink to this headline">¶</a></h4>
<span class="target" id="module-11"></span><dl class="py function">
<dt class="sig sig-object py" id="snpEff_ann.annotate_merged">
<span class="sig-prename descclassname"><span class="pre">snpEff_ann.</span></span><span class="sig-name descname"><span class="pre">annotate_merged</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">datadir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">snpEff</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/snpEff_ann.html#annotate_merged"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#snpEff_ann.annotate_merged" title="Permalink to this definition">¶</a></dt>
<dd><p>Function that calls snpEff to annotated the required
VCF file</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_id</strong> (<em>string</em>) – ID of the patient/sample being analysed</p></li>
<li><p><strong>datadir</strong> (<em>string</em>) – Location of the BAM files</p></li>
<li><p><strong>snpEff</strong> (<em>string</em>) – snpEff executable location</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>returns success or exists</p>
</dd>
<dt class="field-odd">Todo</dt>
<dd class="field-odd"><p>return error</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="snpEff_ann.annotate_pseudo">
<span class="sig-prename descclassname"><span class="pre">snpEff_ann.</span></span><span class="sig-name descname"><span class="pre">annotate_pseudo</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">datadir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">snpEff</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/snpEff_ann.html#annotate_pseudo"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#snpEff_ann.annotate_pseudo" title="Permalink to this definition">¶</a></dt>
<dd><p>Function that calls snpEff to annotated the required
VCF file</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_id</strong> (<em>string</em>) – ID of the patient/sample being analysed</p></li>
<li><p><strong>datadir</strong> (<em>string</em>) – Location of the BAM files</p></li>
<li><p><strong>snpEff</strong> (<em>string</em>) – snpEff executable location</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>returns success or exists</p>
</dd>
<dt class="field-odd">Todo</dt>
<dd class="field-odd"><p>return error</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-variants_freebayes">
<span id="variants-freebayes-py"></span><h4>variants_freebayes.py<a class="headerlink" href="#module-variants_freebayes" title="Permalink to this headline">¶</a></h4>
<span class="target" id="module-12"></span><dl class="py function">
<dt class="sig sig-object py" id="variants_freebayes.edit_freebayes_vcf">
<span class="sig-prename descclassname"><span class="pre">variants_freebayes.</span></span><span class="sig-name descname"><span class="pre">edit_freebayes_vcf</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">directory</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/variants_freebayes.html#edit_freebayes_vcf"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#variants_freebayes.edit_freebayes_vcf" title="Permalink to this definition">¶</a></dt>
<dd><p>Function that removes extra lines in Freebayes generated VCF to allow proper sorting</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_id</strong> (<em>string</em>) – ID of the patient/sample being analysed using GATK</p></li>
<li><p><strong>directory</strong> (<em>string</em>) – Location of the BAM files</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="variants_freebayes.freebayes">
<span class="sig-prename descclassname"><span class="pre">variants_freebayes.</span></span><span class="sig-name descname"><span class="pre">freebayes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">directory</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bed_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">freebayes</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/variants_freebayes.html#freebayes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#variants_freebayes.freebayes" title="Permalink to this definition">¶</a></dt>
<dd><p>Function that calls Freebayes to generate a VCF file</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_id</strong> (<em>string</em>) – ID of the patient/sample being analysed</p></li>
<li><p><strong>directory</strong> (<em>string</em>) – Location of the BAM files</p></li>
<li><p><strong>bed_file</strong> (<em>string</em>) – BED file with regions to be analysed</p></li>
<li><p><strong>reference</strong> (<em>string</em>) – Reference file used in the original alignment</p></li>
<li><p><strong>freebayes</strong> (<em>string</em>) – Location of Freebayes executable</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>returns success or exists</p>
</dd>
<dt class="field-odd">Todo</dt>
<dd class="field-odd"><p>return error</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-13">
<span id="variants-octopus-py"></span><h4>variants_octopus.py<a class="headerlink" href="#module-13" title="Permalink to this headline">¶</a></h4>
<span class="target" id="module-14"></span><dl class="py function">
<dt class="sig sig-object py" id="id0">
<span class="sig-prename descclassname"><span class="pre">variants_freebayes.</span></span><span class="sig-name descname"><span class="pre">edit_freebayes_vcf</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">directory</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/variants_freebayes.html#edit_freebayes_vcf"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#id0" title="Permalink to this definition">¶</a></dt>
<dd><p>Function that removes extra lines in Freebayes generated VCF to allow proper sorting</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_id</strong> (<em>string</em>) – ID of the patient/sample being analysed using GATK</p></li>
<li><p><strong>directory</strong> (<em>string</em>) – Location of the BAM files</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="id3">
<span class="sig-prename descclassname"><span class="pre">variants_freebayes.</span></span><span class="sig-name descname"><span class="pre">freebayes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">directory</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bed_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">freebayes</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/variants_freebayes.html#freebayes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#id3" title="Permalink to this definition">¶</a></dt>
<dd><p>Function that calls Freebayes to generate a VCF file</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_id</strong> (<em>string</em>) – ID of the patient/sample being analysed</p></li>
<li><p><strong>directory</strong> (<em>string</em>) – Location of the BAM files</p></li>
<li><p><strong>bed_file</strong> (<em>string</em>) – BED file with regions to be analysed</p></li>
<li><p><strong>reference</strong> (<em>string</em>) – Reference file used in the original alignment</p></li>
<li><p><strong>freebayes</strong> (<em>string</em>) – Location of Freebayes executable</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>returns success or exists</p>
</dd>
<dt class="field-odd">Todo</dt>
<dd class="field-odd"><p>return error</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-variants_GATK">
<span id="variants-gatk-py"></span><h4>variants_GATK.py<a class="headerlink" href="#module-variants_GATK" title="Permalink to this headline">¶</a></h4>
<span class="target" id="module-15"></span><dl class="py function">
<dt class="sig sig-object py" id="variants_GATK.haplotype_caller">
<span class="sig-prename descclassname"><span class="pre">variants_GATK.</span></span><span class="sig-name descname"><span class="pre">haplotype_caller</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">directory</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bed_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gatk</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/variants_GATK.html#haplotype_caller"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#variants_GATK.haplotype_caller" title="Permalink to this definition">¶</a></dt>
<dd><p>Function that calls GATK’s HaplotypeCaller parameter to generate a list
of raw variants</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_id</strong> (<em>string</em>) – ID of the patient/sample being analysed using GATK</p></li>
<li><p><strong>directory</strong> (<em>string</em>) – Location of the BAM files</p></li>
<li><p><strong>bed_file</strong> (<em>string</em>) – BED file with regions to be analysed</p></li>
<li><p><strong>reference</strong> (<em>string</em>) – Reference file used in the original alignment</p></li>
<li><p><strong>gatk</strong> (<em>string</em>) – GATK jar file location</p></li>
<li><p><strong>pcr</strong> (<em>bool</em>) – If True variant call is made on BWA original’s BAM</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>returns success or exists</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="variants-gatk3-py">
<h4>variants_GATK3.py<a class="headerlink" href="#variants-gatk3-py" title="Permalink to this headline">¶</a></h4>
<span class="target" id="module-16"></span><span class="target" id="module-17"></span><dl class="py function">
<dt class="sig sig-object py" id="id4">
<span class="sig-prename descclassname"><span class="pre">variants_GATK.</span></span><span class="sig-name descname"><span class="pre">haplotype_caller</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">directory</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bed_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gatk</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/variants_GATK.html#haplotype_caller"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#id4" title="Permalink to this definition">¶</a></dt>
<dd><p>Function that calls GATK’s HaplotypeCaller parameter to generate a list
of raw variants</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_id</strong> (<em>string</em>) – ID of the patient/sample being analysed using GATK</p></li>
<li><p><strong>directory</strong> (<em>string</em>) – Location of the BAM files</p></li>
<li><p><strong>bed_file</strong> (<em>string</em>) – BED file with regions to be analysed</p></li>
<li><p><strong>reference</strong> (<em>string</em>) – Reference file used in the original alignment</p></li>
<li><p><strong>gatk</strong> (<em>string</em>) – GATK jar file location</p></li>
<li><p><strong>pcr</strong> (<em>bool</em>) – If True variant call is made on BWA original’s BAM</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>returns success or exists</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-18">
<span id="variants-table-py"></span><h4>variants_table.py<a class="headerlink" href="#module-18" title="Permalink to this headline">¶</a></h4>
<span class="target" id="module-19"></span><dl class="py function">
<dt class="sig sig-object py" id="id5">
<span class="sig-prename descclassname"><span class="pre">variants_GATK.</span></span><span class="sig-name descname"><span class="pre">haplotype_caller</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">directory</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bed_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gatk</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/variants_GATK.html#haplotype_caller"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#id5" title="Permalink to this definition">¶</a></dt>
<dd><p>Function that calls GATK’s HaplotypeCaller parameter to generate a list
of raw variants</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_id</strong> (<em>string</em>) – ID of the patient/sample being analysed using GATK</p></li>
<li><p><strong>directory</strong> (<em>string</em>) – Location of the BAM files</p></li>
<li><p><strong>bed_file</strong> (<em>string</em>) – BED file with regions to be analysed</p></li>
<li><p><strong>reference</strong> (<em>string</em>) – Reference file used in the original alignment</p></li>
<li><p><strong>gatk</strong> (<em>string</em>) – GATK jar file location</p></li>
<li><p><strong>pcr</strong> (<em>bool</em>) – If True variant call is made on BWA original’s BAM</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>returns success or exists</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="vcf-parser-py">
<h4>vcf_parser.py<a class="headerlink" href="#vcf-parser-py" title="Permalink to this headline">¶</a></h4>
</section>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="commands_ngs.html" class="btn btn-neutral float-left" title="Commands" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="code_data_mgmt.html" class="btn btn-neutral float-right" title="Code - Data Management" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2015-2019, Paulo Nuin.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>