<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Running the pipeline at GLS North &mdash; MDL 1 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Running the pipeline at GLS South" href="running_south.html" />
    <link rel="prev" title="NGS pipeline" href="ngs_pipeline.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> MDL
          </a>
              <div class="version">
                Version 1.4.5
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="ngs_pipeline.html">NGS pipeline</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Running the pipeline at GLS North</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#hardware-infrastructure">Hardware infrastructure</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#sequencers">Sequencers</a></li>
<li class="toctree-l3"><a class="reference internal" href="#computers">Computers</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#location-of-programs-files-in-each-machine">Location of programs/files in each machine</a></li>
<li class="toctree-l2"><a class="reference internal" href="#current-status-of-the-automation">Current status of the automation</a></li>
<li class="toctree-l2"><a class="reference internal" href="#manually-setting-up-a-pipeline-run">Manually setting up a pipeline run</a><ul class="simple">
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="running_south.html">Running the pipeline at GLS South</a></li>
<li class="toctree-l1"><a class="reference internal" href="genes.html">Genes</a></li>
<li class="toctree-l1"><a class="reference internal" href="applications.html">Applications</a></li>
<li class="toctree-l1"><a class="reference internal" href="commands_ngs.html">Commands</a></li>
<li class="toctree-l1"><a class="reference internal" href="code.html">Code</a></li>
<li class="toctree-l1"><a class="reference internal" href="code_data_mgmt.html">Code - Data Management</a></li>
<li class="toctree-l1"><a class="reference internal" href="code_webapps.html">Code - Webapps</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">MDL</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
      <li>Running the pipeline at GLS North</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/running_north.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="running-the-pipeline-at-gls-north">
<h1>Running the pipeline at GLS North<a class="headerlink" href="#running-the-pipeline-at-gls-north" title="Permalink to this headline">¶</a></h1>
<section id="hardware-infrastructure">
<h2>Hardware infrastructure<a class="headerlink" href="#hardware-infrastructure" title="Permalink to this headline">¶</a></h2>
<p>Currently GLS North has two sequencers and five computers to run the analysis of NGS samples. Changes will be made when a dedicated task manager is incorporated in the system, but at the moment these are the main actors:</p>
<section id="sequencers">
<h3>Sequencers<a class="headerlink" href="#sequencers" title="Permalink to this headline">¶</a></h3>
<p>The two sequences are:</p>
<ul class="simple">
<li><p>NextSeq</p></li>
<li><p>MiSeq</p></li>
</ul>
<p>Analysis of MiSeq data takes roughly one hour per sample, while NextSeq takes around 4 hours in total.</p>
</section>
<section id="computers">
<h3>Computers<a class="headerlink" href="#computers" title="Permalink to this headline">¶</a></h3>
<p>The computers used for the analysis are:</p>
<ul class="simple">
<li><p>192.168.1.27 - Weinberg (soon to be dedicated to MiSeq)</p></li>
<li><p>192.168.1.169 - Knuth - regular analysis box</p></li>
<li><p>192.168.1.171 - Kimura - regular analysis box</p></li>
<li><p>192.168.1.170 - SequenceKing, bcl2fastq processor</p></li>
<li><p>192.168.1.172 - Nei - regular analysis box</p></li>
<li><p>192.168.1.173 - Felsenstein - regular analysis box</p></li>
<li><p>192.168.1.106 - ProteinKing - Linux box, to be used for peptide analysis</p></li>
</ul>
<p>Another machine, 192.168.1.67, the development box hosts a couple of scripts for automation.</p>
</section>
</section>
<section id="location-of-programs-files-in-each-machine">
<h2>Location of programs/files in each machine<a class="headerlink" href="#location-of-programs-files-in-each-machine" title="Permalink to this headline">¶</a></h2>
<p>All machines (.27, .169, .171, .172, .173) have identical configuration regarding applications and required files. All are stored in <code class="docutils literal notranslate"><span class="pre">/opt</span></code> with this structure:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">BED</span>
<span class="nb">bin</span>
<span class="n">bundle</span>
<span class="n">cancerplus</span>
<span class="n">qualimap</span>
<span class="n">reference</span>
</pre></div>
</div>
<p>where</p>
<ul>
<li><p>BED contains all the <code class="docutils literal notranslate"><span class="pre">BED</span></code> and <code class="docutils literal notranslate"><span class="pre">BAIT</span></code> and some <code class="docutils literal notranslate"><span class="pre">list</span></code> files:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Inherited_Cancer_panel</span><span class="o">.</span><span class="n">picard</span><span class="o">.</span><span class="n">bed</span>
<span class="n">Inherited_Cancer_panel_BED_91122_Target_adjusted_FINAL</span><span class="o">.</span><span class="n">bed</span>
<span class="n">Inherited_Cancer_panel_BED_91122_Target_adjusted_FINAL</span><span class="o">.</span><span class="n">picard</span><span class="o">.</span><span class="n">bed</span>
<span class="n">Inherited_Cancer_panel_BED_91122_Target_adjusted_FINAL_APC</span><span class="o">.</span><span class="n">bed</span>
<span class="n">Inherited_Cancer_panel_BED_91122_Target_adjusted_FINAL_APC</span><span class="o">.</span><span class="n">picard</span><span class="o">.</span><span class="n">bed</span>
<span class="n">Inherited_Cancer_panel_BED_91122_Target_adjusted_FINAL_BRCAcore_6genes</span><span class="o">.</span><span class="n">bed</span>
<span class="n">Inherited_Cancer_panel_BED_91122_Target_adjusted_FINAL_BRCAcore_6genes</span><span class="o">.</span><span class="n">picard</span><span class="o">.</span><span class="n">bed</span>
<span class="n">Inherited_Cancer_panel_BED_91122_Target_adjusted_FINAL_BRCAext_13genes</span><span class="o">.</span><span class="n">bed</span>
<span class="n">Inherited_Cancer_panel_BED_91122_Target_adjusted_FINAL_BRCAext_13genes</span><span class="o">.</span><span class="n">picard</span><span class="o">.</span><span class="n">bed</span>
<span class="n">Inherited_Cancer_panel_BED_91122_Target_adjusted_FINAL_BRCAext_Lynch_17genes</span><span class="o">.</span><span class="n">bed</span>
<span class="n">Inherited_Cancer_panel_BED_91122_Target_adjusted_FINAL_BRCAext_Lynch_17genes</span><span class="o">.</span><span class="n">picard</span><span class="o">.</span><span class="n">bed</span>
<span class="n">Inherited_Cancer_panel_BED_91122_Target_adjusted_FINAL_BRCApanel</span><span class="o">.</span><span class="n">bed</span>
<span class="n">Inherited_Cancer_panel_BED_91122_Target_adjusted_FINAL_BRCApanel</span><span class="o">.</span><span class="n">picard</span><span class="o">.</span><span class="n">bed</span>
<span class="n">Inherited_Cancer_panel_BED_91122_Target_adjusted_FINAL_DNArepair_5genes</span><span class="o">.</span><span class="n">bed</span>
<span class="n">Inherited_Cancer_panel_BED_91122_Target_adjusted_FINAL_DNArepair_5genes</span><span class="o">.</span><span class="n">picard</span><span class="o">.</span><span class="n">bed</span>
<span class="n">Inherited_Cancer_panel_BED_91122_Target_adjusted_FINAL_FA_15genes</span><span class="o">.</span><span class="n">bed</span>
<span class="n">Inherited_Cancer_panel_BED_91122_Target_adjusted_FINAL_FA_15genes</span><span class="o">.</span><span class="n">picard</span><span class="o">.</span><span class="n">bed</span>
<span class="n">Inherited_Cancer_panel_BED_91122_Target_adjusted_FINAL_FA_20genes</span><span class="o">.</span><span class="n">bed</span>
<span class="n">Inherited_Cancer_panel_BED_91122_Target_adjusted_FINAL_FA_20genes</span><span class="o">.</span><span class="n">picard</span><span class="o">.</span><span class="n">bed</span>
<span class="n">Inherited_Cancer_panel_BED_91122_Target_adjusted_FINAL_GIPoly</span><span class="o">.</span><span class="n">bed</span>
<span class="n">Inherited_Cancer_panel_BED_91122_Target_adjusted_FINAL_GIPoly</span><span class="o">.</span><span class="n">picard</span><span class="o">.</span><span class="n">bed</span>
<span class="n">Inherited_Cancer_panel_BED_91122_Target_adjusted_FINAL_HNPCC_4genes</span><span class="o">.</span><span class="n">bed</span>
<span class="n">Inherited_Cancer_panel_BED_91122_Target_adjusted_FINAL_HNPCC_4genes</span><span class="o">.</span><span class="n">picard</span><span class="o">.</span><span class="n">bed</span>
<span class="n">Inherited_Cancer_panel_BED_91122_Target_adjusted_FINAL_HNPCC_4genes_PMS2_1_10</span><span class="o">.</span><span class="n">bed</span>
<span class="n">Inherited_Cancer_panel_BED_91122_Target_adjusted_FINAL_HNPCC_4genes_PMS2_1_10</span><span class="o">.</span><span class="n">picard</span><span class="o">.</span><span class="n">bed</span>
<span class="n">Inherited_Cancer_panel_BED_91122_Target_adjusted_FINAL_HNPCC_PMS2FULL</span><span class="o">.</span><span class="n">bed</span>
<span class="n">Inherited_Cancer_panel_BED_91122_Target_adjusted_FINAL_HNPCC_PMS2FULL</span><span class="o">.</span><span class="n">picard</span><span class="o">.</span><span class="n">bed</span>
<span class="n">Inherited_Cancer_panel_BED_91122_Target_adjusted_FINAL_HNPCCpanel</span><span class="o">.</span><span class="n">bed</span>
<span class="n">Inherited_Cancer_panel_BED_91122_Target_adjusted_FINAL_HNPCCpanel</span><span class="o">.</span><span class="n">picard</span><span class="o">.</span><span class="n">bed</span>
<span class="n">Inherited_Cancer_panel_BED_91122_Target_adjusted_FINAL_PMS2_Ex11_15</span><span class="o">.</span><span class="n">bed</span>
<span class="n">Inherited_Cancer_panel_BED_91122_Target_adjusted_FINAL_PMS2_Ex11_15</span><span class="o">.</span><span class="n">picard</span><span class="o">.</span><span class="n">bed</span>
<span class="n">Inherited_Cancer_panel_BED_91122_Target_adjusted_FINAL_PTEN</span><span class="o">.</span><span class="n">bed</span>
<span class="n">Inherited_Cancer_panel_BED_91122_Target_adjusted_FINAL_PTEN</span><span class="o">.</span><span class="n">picard</span><span class="o">.</span><span class="n">bed</span>
<span class="n">Inherited_Cancer_panel_BED_91122_Target_adjusted_FINAL_Pancreatic_14genes</span><span class="o">.</span><span class="n">bed</span>
<span class="n">Inherited_Cancer_panel_BED_91122_Target_adjusted_FINAL_Pancreatic_14genes</span><span class="o">.</span><span class="n">picard</span><span class="o">.</span><span class="n">bed</span>
<span class="n">Inherited_Cancer_panel_BED_91122_Target_adjusted_FINAL_STK11</span><span class="o">.</span><span class="n">bed</span>
<span class="n">Inherited_Cancer_panel_BED_91122_Target_adjusted_FINAL_STK11</span><span class="o">.</span><span class="n">picard</span><span class="o">.</span><span class="n">bed</span>
<span class="n">Inherited_Cancer_panel_BED_91122_Target_adjusted_FINAL_TP53</span><span class="o">.</span><span class="n">bed</span>
<span class="n">Inherited_Cancer_panel_BED_91122_Target_adjusted_FINAL_TP53</span><span class="o">.</span><span class="n">picard</span><span class="o">.</span><span class="n">bed</span>
<span class="n">Inherited_Cancer_panel_BED_91122_Target_adjusted_FINAL_both</span><span class="o">.</span><span class="n">bed</span>
<span class="n">Inherited_Cancer_panel_BED_91122_Target_adjusted_FINAL_both</span><span class="o">.</span><span class="n">picard</span><span class="o">.</span><span class="n">bed</span>
<span class="n">Inherited_Cancer_panel_FINAL</span><span class="o">-</span><span class="nb">sorted</span><span class="o">.</span><span class="n">bed</span>
<span class="n">Inherited_Cancer_panel_FINAL</span><span class="o">.</span><span class="n">bed</span>
<span class="n">Inherited_Cancer_panel_FINAL</span><span class="o">.</span><span class="n">list</span>
<span class="n">Inherited_Cancer_panel_FINAL</span><span class="o">.</span><span class="n">picard</span><span class="o">.</span><span class="n">bed</span>
</pre></div>
</div>
</li>
<li><p>bin (should) contains all the executable programs</p></li>
<li><p>bundle contains all the GATK related bundle files</p></li>
<li><p>cancerplus is a git repository with the whole pipeline code, ready to run</p></li>
<li><p>qualimap has Qualimap’s executable (Java)</p></li>
<li><p>reference contains all hg19 files and related indexes and other files:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">glist</span><span class="o">-</span><span class="n">hg19</span>
<span class="n">glist</span><span class="o">-</span><span class="n">hg19_2</span>
<span class="n">hg19</span><span class="mf">.1</span><span class="o">.</span><span class="n">bt2</span>
<span class="n">hg19</span><span class="mf">.2</span><span class="o">.</span><span class="n">bt2</span>
<span class="n">hg19</span><span class="mf">.3</span><span class="o">.</span><span class="n">bt2</span>
<span class="n">hg19</span><span class="mf">.4</span><span class="o">.</span><span class="n">bt2</span>
<span class="n">hg19</span><span class="o">.</span><span class="n">dict</span>
<span class="n">hg19</span><span class="o">.</span><span class="n">fasta</span>
<span class="n">hg19</span><span class="o">.</span><span class="n">fasta</span><span class="o">.</span><span class="n">amb</span>
<span class="n">hg19</span><span class="o">.</span><span class="n">fasta</span><span class="o">.</span><span class="n">ann</span>
<span class="n">hg19</span><span class="o">.</span><span class="n">fasta</span><span class="o">.</span><span class="n">bwt</span>
<span class="n">hg19</span><span class="o">.</span><span class="n">fasta</span><span class="o">.</span><span class="n">fai</span>
<span class="n">hg19</span><span class="o">.</span><span class="n">fasta</span><span class="o">.</span><span class="n">pac</span>
<span class="n">hg19</span><span class="o">.</span><span class="n">fasta</span><span class="o">.</span><span class="n">sa</span>
<span class="n">hg19</span><span class="o">.</span><span class="n">rev</span><span class="mf">.1</span><span class="o">.</span><span class="n">bt2</span>
<span class="n">hg19</span><span class="o">.</span><span class="n">rev</span><span class="mf">.2</span><span class="o">.</span><span class="n">bt2</span>
<span class="n">hg19_index</span><span class="mf">.1</span><span class="o">.</span><span class="n">ebwt</span>
<span class="n">hg19_index</span><span class="mf">.2</span><span class="o">.</span><span class="n">ebwt</span>
<span class="n">hg19_index</span><span class="mf">.3</span><span class="o">.</span><span class="n">ebwt</span>
<span class="n">hg19_index</span><span class="mf">.4</span><span class="o">.</span><span class="n">ebwt</span>
<span class="n">hg19_index</span><span class="o">.</span><span class="n">rev</span><span class="mf">.1</span><span class="o">.</span><span class="n">ebwt</span>
<span class="n">hg19_index</span><span class="o">.</span><span class="n">rev</span><span class="mf">.2</span><span class="o">.</span><span class="n">ebwt</span>
</pre></div>
</div>
</li>
</ul>
</section>
<section id="current-status-of-the-automation">
<h2>Current status of the automation<a class="headerlink" href="#current-status-of-the-automation" title="Permalink to this headline">¶</a></h2>
<p>A script <code class="docutils literal notranslate"><span class="pre">watch_bcl.py</span></code> is currently running on SequenceKing (192.168.1.170) and is resposible for the FASTQ generation. MiSeq and NextSeq are already saving BCL files in this machine and conversion is automatic. At the moment, data is being transferred only to 192.168.1.171, more specifically to <code class="docutils literal notranslate"><span class="pre">/Volumes/Bradbury/CancerPlusRuns/&lt;run</span> <span class="pre">ID&gt;/BaseCalls</span></code>.</p>
<p>A script, <code class="docutils literal notranslate"><span class="pre">watch_sequences.py</span></code>, will be set in all analysis boxes and is responsible for automatically starting the pipeline analysis of MiSeq samples. This scripts monitors directories created on <code class="docutils literal notranslate"><span class="pre">/Volumes/&lt;external_storage</span></code> with a <code class="docutils literal notranslate"><span class="pre">Cplus_201*</span></code> tag. When a new folder is found, the script will get the configuration file created by NGSRunSetup.exe</p>
<p>Results are automatically generated on the same directtory <code class="docutils literal notranslate"><span class="pre">/Volumes/Bradbury/CancerPlusRuns/&lt;run</span> <span class="pre">ID&gt;/</span></code> under the <code class="docutils literal notranslate"><span class="pre">BAM</span></code> directory.</p>
</section>
<section id="manually-setting-up-a-pipeline-run">
<h2>Manually setting up a pipeline run<a class="headerlink" href="#manually-setting-up-a-pipeline-run" title="Permalink to this headline">¶</a></h2>
<p>It is pretty simple to setup a run following these instructions:</p>
<ul>
<li><p>scp the <code class="docutils literal notranslate"><span class="pre">run</span> <span class="pre">configuration</span> <span class="pre">file</span></code> created by NGSRunSetup to <code class="docutils literal notranslate"><span class="pre">/Volumes/Bradbury/CancerPlusRuns/&lt;run_ID&gt;</span></code></p></li>
<li><p>ssh into 192.168.1.171 using the <code class="docutils literal notranslate"><span class="pre">pipeline</span></code> user (contact me for password)</p></li>
<li><p>Structure of the configuration file:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">BAIT</span><span class="p">:</span> <span class="o">/</span><span class="n">opt</span><span class="o">/</span><span class="n">BED</span><span class="o">/</span><span class="n">Inherited_Cancer_panel_FINAL</span><span class="o">.</span><span class="n">picard</span><span class="o">.</span><span class="n">bed</span>
<span class="n">BED</span><span class="p">:</span>
  <span class="o">&lt;</span><span class="n">sample</span> <span class="n">ID</span><span class="o">&gt;</span><span class="p">:</span> <span class="o">&lt;</span><span class="n">bed</span> <span class="n">file</span><span class="o">&gt;</span>
  <span class="o">&lt;</span><span class="n">sample</span> <span class="n">ID</span><span class="o">&gt;</span><span class="p">:</span> <span class="o">&lt;</span><span class="n">bed</span> <span class="n">file</span><span class="o">&gt;</span>
  <span class="o">&lt;</span><span class="n">sample</span> <span class="n">ID</span><span class="o">&gt;</span><span class="p">:</span> <span class="o">&lt;</span><span class="n">bed</span> <span class="n">file</span><span class="o">&gt;</span>
  <span class="o">&lt;</span><span class="n">sample</span> <span class="n">ID</span><span class="o">&gt;</span><span class="p">:</span> <span class="o">&lt;</span><span class="n">bed</span> <span class="n">file</span><span class="o">&gt;</span>
  <span class="o">&lt;</span><span class="n">sample</span> <span class="n">ID</span><span class="o">&gt;</span><span class="p">:</span> <span class="o">&lt;</span><span class="n">bed</span> <span class="n">file</span><span class="o">&gt;</span>
  <span class="o">&lt;</span><span class="n">sample</span> <span class="n">ID</span><span class="o">&gt;</span><span class="p">:</span> <span class="o">&lt;</span><span class="n">bed</span> <span class="n">file</span><span class="o">&gt;</span>
  <span class="o">&lt;</span><span class="n">sample</span> <span class="n">ID</span><span class="o">&gt;</span><span class="p">:</span> <span class="o">&lt;</span><span class="n">bed</span> <span class="n">file</span><span class="o">&gt;</span>
  <span class="o">&lt;</span><span class="n">sample</span> <span class="n">ID</span><span class="o">&gt;</span><span class="p">:</span> <span class="o">&lt;</span><span class="n">bed</span> <span class="n">file</span><span class="o">&gt;</span>
  <span class="o">&lt;</span><span class="n">sample</span> <span class="n">ID</span><span class="o">&gt;</span><span class="p">:</span> <span class="o">&lt;</span><span class="n">bed</span> <span class="n">file</span><span class="o">&gt;</span>
  <span class="o">&lt;</span><span class="n">sample</span> <span class="n">ID</span><span class="o">&gt;</span><span class="p">:</span> <span class="o">&lt;</span><span class="n">bed</span> <span class="n">file</span><span class="o">&gt;</span>
  <span class="o">&lt;</span><span class="n">sample</span> <span class="n">ID</span><span class="o">&gt;</span><span class="p">:</span> <span class="o">&lt;</span><span class="n">bed</span> <span class="n">file</span><span class="o">&gt;</span>
  <span class="o">&lt;</span><span class="n">sample</span> <span class="n">ID</span><span class="o">&gt;</span><span class="p">:</span> <span class="o">&lt;</span><span class="n">bed</span> <span class="n">file</span><span class="o">&gt;</span>
  <span class="o">&lt;</span><span class="n">sample</span> <span class="n">ID</span><span class="o">&gt;</span><span class="p">:</span> <span class="o">&lt;</span><span class="n">bed</span> <span class="n">file</span><span class="o">&gt;</span>
  <span class="o">&lt;</span><span class="n">sample</span> <span class="n">ID</span><span class="o">&gt;</span><span class="p">:</span> <span class="o">&lt;</span><span class="n">bed</span> <span class="n">file</span><span class="o">&gt;</span>
  <span class="o">&lt;</span><span class="n">sample</span> <span class="n">ID</span><span class="o">&gt;</span><span class="p">:</span> <span class="o">&lt;</span><span class="n">bed</span> <span class="n">file</span><span class="o">&gt;</span>
  <span class="o">&lt;</span><span class="n">sample</span> <span class="n">ID</span><span class="o">&gt;</span><span class="p">:</span> <span class="o">&lt;</span><span class="n">bed</span> <span class="n">file</span><span class="o">&gt;</span>
  <span class="o">&lt;</span><span class="n">sample</span> <span class="n">ID</span><span class="o">&gt;</span><span class="p">:</span> <span class="o">&lt;</span><span class="n">bed</span> <span class="n">file</span><span class="o">&gt;</span>
  <span class="o">&lt;</span><span class="n">sample</span> <span class="n">ID</span><span class="o">&gt;</span><span class="p">:</span> <span class="o">&lt;</span><span class="n">bed</span> <span class="n">file</span><span class="o">&gt;</span>
  <span class="o">&lt;</span><span class="n">sample</span> <span class="n">ID</span><span class="o">&gt;</span><span class="p">:</span> <span class="o">&lt;</span><span class="n">bed</span> <span class="n">file</span><span class="o">&gt;</span>
<span class="n">Datadir</span><span class="p">:</span> <span class="o">/</span><span class="n">Volumes</span><span class="o">/</span><span class="n">Jupiter</span><span class="o">/</span><span class="n">CancerPlusRuns</span><span class="o">/&lt;</span><span class="n">run</span> <span class="nb">dir</span><span class="o">&gt;</span>
<span class="n">Panel</span><span class="p">:</span> <span class="n">CancerPlus</span>
<span class="n">Reference</span><span class="p">:</span> <span class="o">/</span><span class="n">opt</span><span class="o">/</span><span class="n">reference</span><span class="o">/</span><span class="n">hg19</span><span class="o">.</span><span class="n">fasta</span>
<span class="n">VCF</span><span class="p">:</span> <span class="o">/</span><span class="n">opt</span><span class="o">/</span><span class="n">bundle</span><span class="o">/</span><span class="n">dbsnp_138</span><span class="o">.</span><span class="n">hg19</span><span class="o">.</span><span class="n">vcf</span>
</pre></div>
</div>
</li>
<li><p>&lt;bed file&gt; should have the corresponding BED file to be used for that sample, for instance:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">HNPCC</span><span class="p">:</span> <span class="o">/</span><span class="n">opt</span><span class="o">/</span><span class="n">BED</span><span class="o">/</span><span class="n">Inherited_Cancer_panel_BED_91122_Target_adjusted_FINAL_HNPCC_4genes</span><span class="o">.</span><span class="n">bed</span>
<span class="n">BRCA</span><span class="p">:</span> <span class="o">/</span><span class="n">opt</span><span class="o">/</span><span class="n">BED</span><span class="o">/</span><span class="n">Inherited_Cancer_panel_BED_91122_Target_adjusted_FINAL_BRCApanel</span><span class="o">.</span><span class="n">bed</span>
<span class="n">BRCA</span><span class="o">/</span><span class="n">HNPCC</span><span class="p">:</span> <span class="o">/</span><span class="n">opt</span><span class="o">/</span><span class="n">BED</span><span class="o">/</span><span class="n">Inherited_Cancer_panel_BED_91122_Target_adjusted_FINAL_both</span><span class="o">.</span><span class="n">bed</span>
<span class="n">Cplus</span><span class="p">:</span> <span class="o">/</span><span class="n">opt</span><span class="o">/</span><span class="n">BED</span><span class="o">/</span><span class="n">Inherited_Cancer_panel_FINAL</span><span class="o">.</span><span class="n">bed</span>
<span class="o">...</span>
</pre></div>
</div>
</li>
<li><p>cd to <code class="docutils literal notranslate"><span class="pre">/opt/bin/cancerplus/code/pipeline</span></code></p></li>
<li><p>to start the pipeline enter <code class="docutils literal notranslate"><span class="pre">python</span> <span class="pre">align.py</span> <span class="pre">/Volumes/Bradbury/CancerPlusRuns/&lt;run_ID&gt;/&lt;configuration_file_name&gt;</span></code></p></li>
<li><p>you will have to keep the terminal open for the pipeline to run. To be able to close the terminal window, after changing to the pipeline directory, enter <code class="docutils literal notranslate"><span class="pre">tmux</span></code> in the CLI to open a background processor</p></li>
<li><p>with <code class="docutils literal notranslate"><span class="pre">tmux</span></code> you can close the temrinal window and reattach later: after sshing into the machine, just type <code class="docutils literal notranslate"><span class="pre">tmux</span> <span class="pre">at</span></code>.</p></li>
<li><p>You can also Remote Desktop (tightVNC, VNCviewer, etc) into the machine</p></li>
<li><p>Connect to the selected one and use the <code class="docutils literal notranslate"><span class="pre">Pipeline</span> <span class="pre">Runner</span></code> user and enter the password (contact for password)</p></li>
<li><p>On the terminal enter these commands:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">cd</span> <span class="o">/</span><span class="n">opt</span><span class="o">/</span><span class="n">cancerplus</span><span class="o">/</span><span class="n">code</span><span class="o">/</span><span class="n">pipeline</span>
<span class="o">&gt;</span> <span class="n">tmux</span>
<span class="o">&gt;</span> <span class="n">python</span> <span class="n">align</span><span class="o">.</span><span class="n">py</span> <span class="o">/</span><span class="n">Volumes</span><span class="o">/</span><span class="n">Jupiter</span><span class="o">/</span><span class="n">CancerPlusRuns</span><span class="o">/&lt;</span><span class="n">run</span> <span class="nb">id</span><span class="o">&gt;/&lt;</span><span class="n">configuration</span> <span class="n">file</span> <span class="n">name</span><span class="o">&gt;</span>
</pre></div>
</div>
</li>
</ul>
<p>where the run id is the run you want to analyse, i.e. <code class="docutils literal notranslate"><span class="pre">161207_M02619_0072_000000000-ATMCN_Cplus_2016_NGS_10</span></code>, and the YAML file name for this run is <code class="docutils literal notranslate"><span class="pre">Cplus_2016_NGS_10.yaml</span></code></p>
<p>You should see messages coming on the screen and BWA will start the alignment. If unsuccessful, please email support via SysAid or direct message.</p>
<div class="toctree-wrapper compound">
</div>
<ul class="simple">
<li><p><a class="reference internal" href="genindex.html"><span class="std std-ref">Index</span></a></p></li>
<li><p><a class="reference internal" href="py-modindex.html"><span class="std std-ref">Module Index</span></a></p></li>
<li><p><a class="reference internal" href="search.html"><span class="std std-ref">Search Page</span></a></p></li>
</ul>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="ngs_pipeline.html" class="btn btn-neutral float-left" title="NGS pipeline" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="running_south.html" class="btn btn-neutral float-right" title="Running the pipeline at GLS South" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2015-2019, Paulo Nuin.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>