<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>recalibration &mdash; MDL 1 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html" class="icon icon-home"> MDL
          </a>
              <div class="version">
                Version 1.4.5
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../ngs_pipeline.html">NGS pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="../running_north.html">Running the pipeline at GLS North</a></li>
<li class="toctree-l1"><a class="reference internal" href="../running_south.html">Running the pipeline at GLS South</a></li>
<li class="toctree-l1"><a class="reference internal" href="../genes.html">Genes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../applications.html">Applications</a></li>
<li class="toctree-l1"><a class="reference internal" href="../commands_ngs.html">Commands</a></li>
<li class="toctree-l1"><a class="reference internal" href="../code.html">Code</a></li>
<li class="toctree-l1"><a class="reference internal" href="../code_data_mgmt.html">Code - Data Management</a></li>
<li class="toctree-l1"><a class="reference internal" href="../code_webapps.html">Code - Webapps</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">MDL</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
          <li><a href="index.html">Module code</a> &raquo;</li>
      <li>recalibration</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for recalibration</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">.. module:: recalibration</span>
<span class="sd">    :platform: Any</span>
<span class="sd">    :synopsis: Module that does base quality recalibration</span>
<span class="sd">.. moduleauthor:: Paulo Nuin, January 2016</span>

<span class="sd">&quot;&quot;&quot;</span>

<span class="c1"># pylint: disable-msg=too-many-arguments</span>
<span class="c1"># pylint: disable-msg=line-too-long</span>

<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">subprocess</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">import</span> <span class="nn">graypy</span>

<span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span><span class="n">level</span><span class="o">=</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">)</span>
<span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span>
<span class="n">graypy_handler</span> <span class="o">=</span> <span class="n">graypy</span><span class="o">.</span><span class="n">GELFUDPHandler</span><span class="p">(</span><span class="s1">&#39;192.168.1.169&#39;</span><span class="p">,</span> <span class="mi">12201</span><span class="p">)</span>
<span class="n">logger</span><span class="o">.</span><span class="n">addHandler</span><span class="p">(</span><span class="n">graypy_handler</span><span class="p">)</span>


<div class="viewcode-block" id="base_recal1"><a class="viewcode-back" href="../code.html#recalibration.base_recal1">[docs]</a><span class="k">def</span> <span class="nf">base_recal1</span><span class="p">(</span><span class="n">sample_id</span><span class="p">,</span> <span class="n">directory</span><span class="p">,</span> <span class="n">bed_file</span><span class="p">,</span> <span class="n">vcf_file</span><span class="p">,</span> <span class="n">reference</span><span class="p">,</span> <span class="n">gatk</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Function that does the first step of base recalibration, creating a</span>
<span class="sd">    recalibration data table</span>

<span class="sd">    :param sample_id: ID of the patient/sample being analysed</span>
<span class="sd">    :param directory: Location of the BAM files</span>
<span class="sd">    :param bed_file: BED file with regions to be analysed</span>
<span class="sd">    :param vcf_file: VCF file of known regions of variants</span>
<span class="sd">    :param reference: Reference file used in the original alignment</span>
<span class="sd">    :param gatk: GATK jar file location</span>

<span class="sd">    :type sample_id: string</span>
<span class="sd">    :type directory: string</span>
<span class="sd">    :type bed_file: string</span>
<span class="sd">    :type vcf_file: string</span>
<span class="sd">    :type reference: string</span>
<span class="sd">    :type gatk: string</span>

<span class="sd">    :return: returns success or exists</span>

<span class="sd">    :todo: return error</span>
<span class="sd">    :todo: fix argument</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">argument</span> <span class="o">=</span> <span class="n">directory</span> <span class="o">+</span> <span class="s1">&#39;/BAM/&#39;</span> <span class="o">+</span> <span class="n">sample_id</span> <span class="o">+</span> <span class="s1">&#39;/BAM/&#39;</span> <span class="o">+</span> <span class="n">sample_id</span>
    <span class="n">argument2</span> <span class="o">=</span> <span class="n">directory</span> <span class="o">+</span> <span class="s1">&#39;/BAM/&#39;</span> <span class="o">+</span> <span class="n">sample_id</span> <span class="o">+</span> <span class="s1">&#39;/BAM/&#39;</span>

    <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">argument2</span> <span class="o">+</span> <span class="s1">&#39;/recal_data.table&#39;</span><span class="p">):</span>
        <span class="c1"># log_to_kafka(&#39;Recalibration list exists &#39; + argument + &#39; &#39; + MY_HOSTNAME)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Recalibration list exists &#39;</span> <span class="o">+</span> <span class="n">sample_id</span><span class="p">)</span>
        <span class="k">return</span> <span class="s1">&#39;exists&#39;</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Starting step one of base recalibration &#39;</span> <span class="o">+</span> <span class="n">sample_id</span><span class="p">)</span>
    <span class="n">GATK_string</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="si">%s</span><span class="s1"> BaseRecalibrator -R </span><span class="si">%s</span><span class="s1">  -I </span><span class="si">%s</span><span class="s1">.good.bam --known-sites </span><span class="si">%s</span><span class="s1"> -O </span><span class="si">%s</span><span class="s1">/recal_data.table -L </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">gatk</span><span class="p">,</span> <span class="n">reference</span><span class="p">,</span> <span class="n">argument</span><span class="p">,</span> <span class="n">vcf_file</span><span class="p">,</span> <span class="n">argument2</span><span class="p">,</span> <span class="n">bed_file</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Command &#39;</span> <span class="o">+</span> <span class="n">GATK_string</span> <span class="o">+</span>  <span class="s1">&#39; &#39;</span> <span class="o">+</span> <span class="n">sample_id</span><span class="p">)</span>
    <span class="n">proc</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">Popen</span><span class="p">(</span><span class="n">GATK_string</span><span class="p">,</span> <span class="n">shell</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">stdout</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span> <span class="n">stderr</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">)</span>
    <span class="k">while</span> <span class="kc">True</span><span class="p">:</span>
        <span class="n">output</span> <span class="o">=</span> <span class="n">proc</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">readline</span><span class="p">()</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">output</span> <span class="o">==</span> <span class="sa">b</span><span class="s1">&#39;&#39;</span><span class="p">:</span>
            <span class="k">break</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="n">output</span><span class="o">.</span><span class="n">decode</span><span class="p">(</span><span class="s1">&#39;utf-8&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span> <span class="o">+</span> <span class="n">sample_id</span><span class="p">)</span>
    <span class="n">proc</span><span class="o">.</span><span class="n">wait</span><span class="p">()</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Recalibration step one completed successfully &#39;</span> <span class="o">+</span> <span class="n">sample_id</span><span class="p">)</span>
    <span class="k">return</span> <span class="s1">&#39;success&#39;</span></div>


<div class="viewcode-block" id="base_recal2"><a class="viewcode-back" href="../code.html#recalibration.base_recal2">[docs]</a><span class="k">def</span> <span class="nf">base_recal2</span><span class="p">(</span><span class="n">sample_id</span><span class="p">,</span> <span class="n">directory</span><span class="p">,</span> <span class="n">vcf_file</span><span class="p">,</span> <span class="n">reference</span><span class="p">,</span> <span class="n">gatk</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Function that does the second step of the base recalibration</span>
<span class="sd">    process, generating the post-recalibration table</span>

<span class="sd">    :param sample_id: ID of the patient/sample being analysed using GATK</span>
<span class="sd">    :param directory: Location of the BAM files</span>
<span class="sd">    :param vcf_file: VCF file of known regions of variants</span>
<span class="sd">    :param reference: Reference file used in the original alignment</span>
<span class="sd">    :param gatk: GATK jar file location</span>

<span class="sd">    :type sample_id: string</span>
<span class="sd">    :type directory: string</span>
<span class="sd">    :type vcf_file: string</span>
<span class="sd">    :type reference: string</span>
<span class="sd">    :type gatk: string</span>

<span class="sd">    :todo: fix argument</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">argument</span> <span class="o">=</span> <span class="n">directory</span> <span class="o">+</span> <span class="s1">&#39;/BAM/&#39;</span> <span class="o">+</span> <span class="n">sample_id</span> <span class="o">+</span> <span class="s1">&#39;/BAM/&#39;</span> <span class="o">+</span> <span class="n">sample_id</span>
    <span class="n">argument2</span> <span class="o">=</span> <span class="n">directory</span> <span class="o">+</span> <span class="s1">&#39;/BAM/&#39;</span> <span class="o">+</span> <span class="n">sample_id</span> <span class="o">+</span> <span class="s1">&#39;/BAM/&#39;</span>

    <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">argument2</span> <span class="o">+</span> <span class="s1">&#39;/post_recal_data.table&#39;</span><span class="p">):</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Post recalibration list exists &#39;</span> <span class="o">+</span> <span class="n">sample_id</span><span class="p">)</span>
        <span class="k">return</span> <span class="s1">&#39;exists&#39;</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Starting recalibration step two &#39;</span> <span class="o">+</span> <span class="n">sample_id</span><span class="p">)</span>
    <span class="n">GATK_string</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="si">%s</span><span class="s1"> ApplyBQSR -R </span><span class="si">%s</span><span class="s1"> -I </span><span class="si">%s</span><span class="s1">.good.bam  </span><span class="si">%s</span><span class="s1">/recal_data.table -O </span><span class="si">%s</span><span class="s1">/post_recal_data.table&#39;</span> <span class="o">%</span> <span class="p">(</span>
        <span class="n">gatk</span><span class="p">,</span> <span class="n">reference</span><span class="p">,</span> <span class="n">argument</span><span class="p">,</span> <span class="n">argument2</span><span class="p">,</span> <span class="n">argument2</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Command &#39;</span> <span class="o">+</span> <span class="n">GATK_string</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span> <span class="o">+</span> <span class="n">sample_id</span><span class="p">)</span>
    <span class="n">proc</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">Popen</span><span class="p">(</span><span class="n">GATK_string</span><span class="p">,</span> <span class="n">shell</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">stdout</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span> <span class="n">stderr</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">)</span>
    <span class="k">while</span> <span class="kc">True</span><span class="p">:</span>
        <span class="n">output</span> <span class="o">=</span> <span class="n">proc</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">readline</span><span class="p">()</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">output</span> <span class="o">==</span> <span class="sa">b</span><span class="s1">&#39;&#39;</span><span class="p">:</span>
            <span class="k">break</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="n">output</span><span class="o">.</span><span class="n">decode</span><span class="p">(</span><span class="s1">&#39;utf-8&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span> <span class="o">+</span> <span class="n">sample_id</span><span class="p">)</span>
    <span class="n">proc</span><span class="o">.</span><span class="n">wait</span><span class="p">()</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Recalibration step two completed successfully &#39;</span> <span class="o">+</span> <span class="n">sample_id</span><span class="p">)</span>
    <span class="k">return</span> <span class="s1">&#39;success&#39;</span></div>


<div class="viewcode-block" id="recalibrate"><a class="viewcode-back" href="../code.html#recalibration.recalibrate">[docs]</a><span class="k">def</span> <span class="nf">recalibrate</span><span class="p">(</span><span class="n">sample_id</span><span class="p">,</span> <span class="n">directory</span><span class="p">,</span> <span class="n">reference</span><span class="p">,</span> <span class="n">gatk</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Function that performs the third step of the base recalibration process,</span>
<span class="sd">    generating the final BAM file. *.recal_reads.bam</span>

<span class="sd">    :param sample_id: ID of the patient/sample being analysed using GATK</span>
<span class="sd">    :param directory: Location of the BAM files</span>
<span class="sd">    :param reference: Reference file used in the original alignment</span>
<span class="sd">    :param gatk: GATK jar file location</span>

<span class="sd">    :type sample_id: string</span>
<span class="sd">    :type directory: string</span>
<span class="sd">    :type reference: string</span>
<span class="sd">    :type gatk: string</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">argument</span> <span class="o">=</span> <span class="n">directory</span> <span class="o">+</span> <span class="s1">&#39;/BAM/&#39;</span> <span class="o">+</span> <span class="n">sample_id</span> <span class="o">+</span> <span class="s1">&#39;/BAM/&#39;</span> <span class="o">+</span> <span class="n">sample_id</span>
    <span class="n">argument2</span> <span class="o">=</span> <span class="n">directory</span> <span class="o">+</span> <span class="s1">&#39;/BAM/&#39;</span> <span class="o">+</span> <span class="n">sample_id</span> <span class="o">+</span> <span class="s1">&#39;/BAM/&#39;</span>

    <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">argument</span> <span class="o">+</span> <span class="s1">&#39;.recal_reads.bam&#39;</span><span class="p">):</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Recalibrated BAM exists &#39;</span> <span class="o">+</span> <span class="n">sample_id</span><span class="p">)</span>
        <span class="k">return</span> <span class="s1">&#39;exists&#39;</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Starting recalibration &#39;</span> <span class="o">+</span> <span class="n">sample_id</span><span class="p">)</span>
    <span class="n">GATK_string</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="si">%s</span><span class="s1"> ApplyBQSR -R </span><span class="si">%s</span><span class="s1"> -I </span><span class="si">%s</span><span class="s1">.good.bam --bqsr-recal-file </span><span class="si">%s</span><span class="s1">/recal_data.table -O </span><span class="si">%s</span><span class="s1">.recal_reads.bam&#39;</span> <span class="o">%</span> <span class="p">(</span>
        <span class="n">gatk</span><span class="p">,</span> <span class="n">reference</span><span class="p">,</span> <span class="n">argument</span><span class="p">,</span> <span class="n">argument2</span><span class="p">,</span> <span class="n">argument</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Command &#39;</span> <span class="o">+</span> <span class="n">GATK_string</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span> <span class="o">+</span> <span class="n">sample_id</span><span class="p">)</span>
    <span class="n">proc</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">Popen</span><span class="p">(</span><span class="n">GATK_string</span><span class="p">,</span> <span class="n">shell</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                            <span class="n">stdout</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span> <span class="n">stderr</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">)</span>
    <span class="k">while</span> <span class="kc">True</span><span class="p">:</span>
        <span class="n">output</span> <span class="o">=</span> <span class="n">proc</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">readline</span><span class="p">()</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">output</span> <span class="o">==</span> <span class="sa">b</span><span class="s1">&#39;&#39;</span><span class="p">:</span>
            <span class="k">break</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="n">output</span><span class="o">.</span><span class="n">decode</span><span class="p">(</span><span class="s1">&#39;utf-8&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span> <span class="o">+</span> <span class="n">sample_id</span><span class="p">)</span>
    <span class="n">proc</span><span class="o">.</span><span class="n">wait</span><span class="p">()</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Recalibration completed &#39;</span> <span class="o">+</span> <span class="n">sample_id</span><span class="p">)</span>
    <span class="k">return</span> <span class="s1">&#39;success&#39;</span></div>


<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s1">&#39;__main__&#39;</span><span class="p">:</span>

    <span class="n">data_directory</span> <span class="o">=</span> <span class="s1">&#39;/Users/nuin/Projects/Data/Test_dataset/&#39;</span>
    <span class="n">sample_id</span> <span class="o">=</span> <span class="s1">&#39;NA12877_1&#39;</span>
    <span class="n">bed_file</span> <span class="o">=</span> <span class="s1">&#39;/opt/BED/Inherited_Cancer_panel_FINAL.bed&#39;</span>
    <span class="n">vcf_file</span> <span class="o">=</span> <span class="s1">&#39;/opt/bundle/dbsnp_138.hg19.vcf&#39;</span>
    <span class="n">reference</span> <span class="o">=</span> <span class="s1">&#39;/opt/reference/hg19.fasta&#39;</span>
    <span class="n">gatk</span> <span class="o">=</span> <span class="s1">&#39;java -jar /usr/local/bin/GenomeAnalysisTK.jar&#39;</span>

    <span class="n">base_recal1</span><span class="p">(</span><span class="n">sample_id</span><span class="p">,</span> <span class="n">data_directory</span><span class="p">,</span> <span class="n">bed_file</span><span class="p">,</span> <span class="n">vcf_file</span><span class="p">,</span> <span class="n">reference</span><span class="p">,</span> <span class="n">gatk</span><span class="p">)</span>
    <span class="n">base_recal2</span><span class="p">(</span><span class="n">sample_id</span><span class="p">,</span> <span class="n">data_directory</span><span class="p">,</span> <span class="n">vcf_file</span><span class="p">,</span> <span class="n">reference</span><span class="p">,</span> <span class="n">gatk</span><span class="p">)</span>
    <span class="n">recalibrate</span><span class="p">(</span><span class="n">sample_id</span><span class="p">,</span> <span class="n">data_directory</span><span class="p">,</span> <span class="n">reference</span><span class="p">,</span> <span class="n">gatk</span><span class="p">)</span>

<span class="c1"># Later</span>
<span class="c1"># java -jar /usr/local/bin/GenomeAnalysisTK.jar -T AnalyzeCovariates -R  ~/New_projects/reference/hg19.fasta -before recal_data.table -after post_recal_data.table -csv report.csv</span>

<span class="c1"># Rscript ~/New_projects/bundle/BQSR.R report.csv recal_data.table test.pdf</span>
</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2015-2019, Paulo Nuin.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>